src/aosm/azext_aosm/common/registry.py

```diff
@@ -0,0 +1,497 @@
+            registry_name = parts[0]
+            registry_namespace = parts[1] if len(parts) > 1 else None
+
+            acr_match = re.match(ACR_REGISTRY_NAME_PATTERN, registry_name)
+
+            if registry_name not in registries:
+                if acr_match:
+                    registries[registry_name] = AzureContainerRegistry(registry_name)
+                else:
+                    registries[registry_name] = UniversalRegistry(registry_name)
+
+                registry_list.append(registries[registry_name])
+
+            if registry_namespace:
+                # Make sure that the namespace ends with a slash
+                if not registry_namespace.endswith("/"):
+                    registry_namespace += "/"
+                registries[registry_name].add_namespace(registry_namespace)
+            else:
+                registries[registry_name].add_namespace("")
+
+        return registry_list
+
+    def get_registry_list(self) -> List[ContainerRegistry]:
+        """
+        Get the list of registries.
+
+        :return: The list of registries
+        """
+        return self.registry_list
+
+    def _get_registries_for_images(
+        self,
+    ) -> Dict[Tuple[str, str], Tuple["AzureContainerRegistry", str]]:
+        """
+        Cycle through all available registries to find images in each.
+
+        :return: A dictionary where keys are (image, version) and values are (registry, namespace)
+        """
+        registry_for_image = {}
+
+        for registry in self.registry_list:
+
+            # We can only do this for ACRs because docker does not have this functionality.
+            if isinstance(registry, AzureContainerRegistry):
+                registry_for_image.update(registry.get_images_in_registry())
+
+        logger.debug("Images found in the ACR registires: %s", registry_for_image)
+        return registry_for_image
+
+    def find_registry_for_image(
+        self, image, version
+    ) -> Union[Tuple["ContainerRegistry", str], Tuple[None, None]]:
+        """
+        Find the registry and namespace for a given image and version.
+
+        :param image: The image to find the registry for
+            This should be stripped of all namespace (e.g nginx, not samples/nginx)
+        :param version: The version of the image
+        :return: The registry and namespace for the image
+        """
+
+        # If we already found this image, version combination in the available ACRs,
+        # return the registry and namespace.
+        if (image, version) in self.registry_for_image:
+            return self.registry_for_image[(image, version)]
+
+        # If we didn't find the image in the ACRs, cycle through the Universal Registries
+        for registry in self.registry_list:
+            if isinstance(registry, UniversalRegistry):
+                image_registry, namespace = registry.find_image(image, version)
+                if image_registry and namespace is not None:
```

Comment:
I'm not sure what your intended check here was. Did you account for precedence? https://docs.python.org/3/reference/expressions.html#operator-precedence

If you did, and this does work, I'd still be tempted either to add a comment or to make the code explicit by using `(` `)` to group what you want. Otherwise future readers might wonder if there's a bug here (like I did :-) )